File: implementations/c/rcl_lexer.c
#include "rcl_internal.h"

#include <stdlib.h>
#include <string.h>

static char cur(const Lexer *lexer) { return lexer->input[lexer->pos]; }
static int eof(const Lexer *lexer) { return lexer->input[lexer->pos] == '\0'; }

static void advance(Lexer *lexer) {
  if (eof(lexer)) return;
  if (cur(lexer) == '\n') {
    lexer->line++;
    lexer->column = 1;
  } else {
    lexer->column++;
  }
  lexer->pos++;
}

static void skip(Lexer *lexer) {
  while (!eof(lexer)) {
    if (cur(lexer) == '#') {
      while (!eof(lexer) && cur(lexer) != '\n') advance(lexer);
      continue;
    }
    if (cur(lexer) == ' ' || cur(lexer) == '\t' || cur(lexer) == '\r' || cur(lexer) == '\n') {
      advance(lexer);
      continue;
    }
    break;
  }
}

void lexer_init(Lexer *lexer, const char *text) {
  lexer->input = text;
  lexer->pos = 0;
  lexer->line = 1;
  lexer->column = 1;
}

void token_free(Token *token) {
  free(token->lexeme);
  token->lexeme = NULL;
}

static int token_simple(Lexer *lexer, Token *token, TokenKind kind, char ch) {
  token->kind = kind;
  token->line = lexer->line;
  token->column = lexer->column;
  token->lexeme = rcl_strndup(&ch, 1);
  advance(lexer);
  return token->lexeme != NULL;
}

static int lex_identifier(Lexer *lexer, Token *token) {
  size_t start = lexer->pos, line = lexer->line, col = lexer->column;
  while (!eof(lexer) && rcl_is_identifier_char(cur(lexer))) advance(lexer);
  token->lexeme = rcl_strndup(lexer->input + start, lexer->pos - start);
  token->line = line;
  token->column = col;
  if (token->lexeme == NULL) return 0;
  if (strcmp(token->lexeme, "do") == 0) token->kind = TOK_DO;
  else if (strcmp(token->lexeme, "end") == 0) token->kind = TOK_END;
  else token->kind = TOK_IDENTIFIER;
  return 1;
}

static int lex_number(Lexer *lexer, Token *token) {
  size_t start = lexer->pos, line = lexer->line, col = lexer->column;
  if (cur(lexer) == '-') advance(lexer);
  while (!eof(lexer) && cur(lexer) >= '0' && cur(lexer) <= '9') advance(lexer);
  if (cur(lexer) == '.') {
    advance(lexer);
    if (cur(lexer) < '0' || cur(lexer) > '9') return 0;
    while (!eof(lexer) && cur(lexer) >= '0' && cur(lexer) <= '9') advance(lexer);
  }
  token->kind = TOK_NUMBER;
  token->line = line;
  token->column = col;
  token->lexeme = rcl_strndup(lexer->input + start, lexer->pos - start);
  return token->lexeme != NULL;
}

static int lex_string(Lexer *lexer, Token *token, RclError *error) {
  size_t line = lexer->line, col = lexer->column, cap = 32, len = 0;
  char *out = (char *)malloc(cap);
  if (out == NULL) return 0;
  advance(lexer);
  while (!eof(lexer) && cur(lexer) != '"') {
    char ch = cur(lexer);
    if (ch == '\\') {
      advance(lexer);
      if (eof(lexer)) {
        free(out);
        rcl_set_error(error, "unterminated string", line, col);
        return -1;
      }
      ch = cur(lexer);
      if (ch == '"') ch = '"';
      else if (ch == '\\') ch = '\\';
      else if (ch == 'n') ch = '\n';
      else if (ch == 't') ch = '\t';
      else {
        free(out);
        rcl_set_error(error, "invalid escape", lexer->line, lexer->column);
        return -1;
      }
    }
    if (len + 2 > cap) {
      cap *= 2;
      out = (char *)realloc(out, cap);
      if (out == NULL) return 0;
    }
    out[len++] = ch;
    advance(lexer);
  }
  if (eof(lexer)) {
    free(out);
    rcl_set_error(error, "unterminated string", line, col);
    return -1;
  }
  advance(lexer);
  out[len] = '\0';
  token->kind = TOK_STRING;
  token->line = line;
  token->column = col;
  token->lexeme = out;
  return 1;
}

int lexer_next(Lexer *lexer, Token *token, RclError *error) {
  skip(lexer);
  token->lexeme = NULL;
  if (eof(lexer)) {
    token->kind = TOK_EOF;
    token->line = lexer->line;
    token->column = lexer->column;
    return 1;
  }
  if (rcl_is_identifier_start(cur(lexer))) return lex_identifier(lexer, token);
  if (cur(lexer) == '-' || (cur(lexer) >= '0' && cur(lexer) <= '9')) return lex_number(lexer, token);
  if (cur(lexer) == '"') return lex_string(lexer, token, error);
  if (cur(lexer) == '\'') {
    rcl_set_error(error, "single-quoted string usage", lexer->line, lexer->column);
    return -1;
  }
  if (cur(lexer) == '=') return token_simple(lexer, token, TOK_EQUAL, '=');
  if (cur(lexer) == ',') return token_simple(lexer, token, TOK_COMMA, ',');
  if (cur(lexer) == '.') return token_simple(lexer, token, TOK_DOT, '.');
  if (cur(lexer) == '[') return token_simple(lexer, token, TOK_LBRACKET, '[');
  if (cur(lexer) == ']') return token_simple(lexer, token, TOK_RBRACKET, ']');
  rcl_set_error(error, "unexpected character", lexer->line, lexer->column);
  return -1;
}

File: implementations/php/src/RCLLexer.php
<?php

final class RCLLexer {
  private int $i = 0; private int $line = 1; private int $col = 1;
  public function __construct(private string $s) {}
  public function next(): array {
    $this->skip();
    if ($this->i >= strlen($this->s)) return ['type' => 'eof', 'value' => null, 'line' => $this->line, 'col' => $this->col];
    $c = $this->s[$this->i]; $l = $this->line; $p = $this->col;
    if (ctype_alpha($c) || $c === '_') return $this->ident();
    if ($c === '-' || ctype_digit($c)) return $this->number();
    if ($c === '"') return $this->str();
    if ($c === "'") $this->err('single-quoted string usage', $l, $p);
    if ($c === '/' && ($this->s[$this->i + 1] ?? '') === '/') $this->err('unexpected character', $l, $p);
    if (str_contains('=.,[]', $c)) { $this->adv(); return ['type' => $c, 'value' => $c, 'line' => $l, 'col' => $p]; }
    $this->err('unexpected character', $l, $p);
  }
  private function skip(): void {
    while ($this->i < strlen($this->s)) {
      $c = $this->s[$this->i];
      if ($c === '#') { while ($this->i < strlen($this->s) && $this->s[$this->i] !== "\n") $this->adv(); continue; }
      if ($c === ' ' || $c === "\t" || $c === "\r" || $c === "\n") { $this->adv(); continue; }
      break;
    }
  }
  private function ident(): array {
    $l = $this->line; $p = $this->col; $v = '';
    while ($this->i < strlen($this->s) && preg_match('/[A-Za-z0-9_]/', $this->s[$this->i])) { $v .= $this->s[$this->i]; $this->adv(); }
    $t = match ($v) { 'do' => 'do', 'end' => 'end', 'true' => 'bool', 'false' => 'bool', default => 'id' };
    $val = $t === 'bool' ? $v === 'true' : $v;
    return ['type' => $t, 'value' => $val, 'line' => $l, 'col' => $p];
  }
  private function number(): array {
    $l = $this->line; $p = $this->col; $n = ''; $dot = false;
    if ($this->s[$this->i] === '-') { $n .= '-'; $this->adv(); }
    if (!ctype_digit($this->s[$this->i] ?? '')) $this->err('unexpected token', $l, $p);
    while ($this->i < strlen($this->s)) {
      $c = $this->s[$this->i];
      if (ctype_digit($c)) { $n .= $c; $this->adv(); continue; }
      if ($c === '.' && !$dot) { $dot = true; $n .= '.'; $this->adv(); if (!ctype_digit($this->s[$this->i] ?? '')) $this->err('unexpected token', $l, $p); continue; }
      break;
    }
    return ['type' => 'number', 'value' => $dot ? (float)$n : (int)$n, 'raw' => $n, 'line' => $l, 'col' => $p];
  }
  private function str(): array {
    $l = $this->line; $p = $this->col; $this->adv(); $v = '';
    while ($this->i < strlen($this->s)) {
      $c = $this->s[$this->i];
      if ($c === '"') { $this->adv(); return ['type' => 'string', 'value' => $v, 'line' => $l, 'col' => $p]; }
      if ($c === '\\') {
        $e = $this->s[$this->i + 1] ?? null;
        if ($e === null) $this->err('unterminated string', $l, $p);
        $m = ['"' => '"', '\\' => '\\', 'n' => "\n", 't' => "\t"];
        if (!isset($m[$e])) $this->err('invalid escape', $this->line, $this->col);
        $v .= $m[$e]; $this->adv(); $this->adv(); continue;
      }
      if ($c === "\n") $this->err('unterminated string', $l, $p);
      $v .= $c; $this->adv();
    }
    $this->err('unterminated string', $l, $p);
  }
  private function adv(): void { if ($this->s[$this->i++] === "\n") { $this->line++; $this->col = 1; } else $this->col++; }
  private function err(string $m, int $l, int $c): never { throw new RCLParseError("line $l, column $c: $m"); }
}
